Updated and Shortened Grid Syntax (#49)
microsoft/microsoft-ui-xaml-specs: active/gridsyntax/GridSyntaxSpec.md

# Examples
### New succinct syntax
The code below has the same functionality as the code shown above with the original syntax. It creates a grid and defines five different rows and columns, each with their own specific height/width, and adds them to the Grid. 
```xml
<Grid ColumnDefinitions="1*, 2*, Auto, *, 300"
      RowDefinitions="1*, Auto, 25, 14, 20">
</Grid>
```

### New succinct syntax for more complex types
The code below shows an example of how the syntax would be used for collection-typed properties that have more complex value types.

This shows how Point objects will be delimited by a comma, but indvidual values will be encased in single quotes or double quotes, whichever is not used to enclose the attribute value.
```xml
<Foo PointList=" '1,2', '9,4', '2,6'"/>
```

> This shows how String values will be supported, even if they include commas or quotes inside of them. Attribute values (i.e `Words` in this example) can be enclosed with either ' or ", so whichever quotation mark is not used can be employed within the value itself, and does not need to be escaped as an XML character entity. The collection shown below has 4 elements, and 'hello, world!' is one element.

Reviewer: ```suggestion
This shows how String values will be supported, even if they include commas or quotes inside of them. Following standard XML syntax,  you can enclose portions within attribute values (e.g. `hello` in this example) with opposite quotes. So for example if the attribute value uses double quotes, you can use single quotes within the attribute value. The collection shown below has 4 elements, and 'hello, world!' is one element.
```

    <Grid.RowDefinitions>
          <RowDefinition>1*</RowDefinition>
          <RowDefinition>Auto</RowDefinition>
          <RowDefinition>25</RowDefinition>
          <RowDefinition>14</RowDefinition>
          <RowDefinition>20</RowDefinition>
    </Grid.RowDefinitions>
</Grid>
```
<!-- Use this section to explain the features of the API, showing
example code with each description. The general format is: 
  feature explanation,
  example code
  feature explanation,
  example code
  etc.-->
  
<!-- Code samples should be in C# and/or C++/WinRT -->

<!-- As an example of this section, see the Examples section for the PasswordBox control 
(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->

# Remarks
<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->
### Syntax Details and Corner Cases

> The new succinct syntax is formed by assigning a collection-type property to a collection of values. This collection of values is encased between a set of double quotes. String values, as well as any other more complex type that may require commas within it (i.e. Point objects, sets), are encased between a set of single quotes. This ensures that string values may contain commas, and Points, sets, or any other kind of nested collection object may be supported. To use a single quote within a complex value that is encased in single quotes (i.e., to put a single quote or apostrophe inside of a string value), the HTML entity (&apos;) should be used to escape that character.

Reviewer: ```suggestion
The new succinct syntax is formed by assigning a collection of values to a read-only collection-type property. This collection of values is encased between a set of quotes. String values, as well as any other more complex types that may require commas within it (i.e. Point objects, sets), are encased between a set of opposite quotes. This ensures that string values may contain commas, and Points, sets, or any other kind of nested collection object may be supported.
```

modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

# Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->
A new XAML language feature that allows the initialization of collection-type properties (including read-only properties) using element attribute syntax. We will also be assigning properties within Grid to make sure that this syntax works properly specifically for Grid's use cases.

1. When the parser encounters an element attribute that is assigning to a collection-type property, it recognizes that initialization of a collection from a string is being requested.
2. The parser tokenizes the attribute value (a comma-delimited string) into a series of substrings, each of which represents the initialization value of a single element of the collection

> 3. The parser will perform the usual "initialize an instance of a type from a string" operation.

Author: I put in a reference to section 6.6.4 on Value Creation from Attribute Text - hope that's the correct section, but we can discuss more at the spec review.

In order to support this language feature in Grid, a mechanism is needed that allows ColumnDefinition and RowDefinition objects to be createable from their Width/Height properties, respectively. To achieve this, the content property of ColumnDefinition is set to the Width property, and the content property of RowDefinition is set to the Height property.This will be used to make the new syntax (see below) fully functional.

# Examples
### New succinct syntax
The code below has the same functionality as the code shown above with the original syntax. It creates a grid and defines five different rows and columns, each with their own specific height/width, and adds them to the Grid. 
```xml
<Grid ColumnDefinitions="1*, 2*, Auto, *, 300"
      RowDefinitions="1*, Auto, 25, 14, 20">
</Grid>
```

### New succinct syntax for more complex types
The code below shows an example of how the syntax would be used for collection-typed properties that have more complex value types.

This shows how Point objects will be delimited by a comma, but indvidual values will be encased in single quotes or double quotes, whichever is not used to enclose the attribute value.
```xml

> <Foo PointList=" '1,2', '9,4', '2,6'"/>

Author: @MikeHillberg we previously looked at ColorWheel for Silverlight and its Colors property, is that what you're referring to? (Sorry for the later responses! Just getting to this now.)

> <Foo PointList=" '1,2', '9,4', '2,6'"/>

Reviewer: Yes, that one's more real-world than <Foo> :)
Although you have other examples below too.

> <Foo PointList=" '1,2', '9,4', '2,6'"/>

Author: Just changed it 👍

    ...
  </Grid.RowDefinitions>
</Grid>
```
Grid's usability, intuitiveness, and developer satsifaction will be improved upon by utilizing a more succinct syntax, as will be described throughout this spec.
<!-- Use this section to provide background context for the new API(s) 
in this spec. -->

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

# Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

> A new XAML language feature that allows read-only collection-type properties to be created by assigning the CreateFromString attribute. The CreateFromString attribute will allow these read-only collection-type properties to be defined/created from a comma-delimited string. In terms of Grid, this means that the ColumnDefinitions and RowDefinitions properties are able to be defined by providing a comma-delimited string of ColumnDefinitions (Width values) and a comma-delimited string of RowDefinitions (Height values), respectively. However, this feature is also applicable to other scenarios within the XAML language where collection-typed read-only properties are created (see Remarks for examples of such scenarios). 

Reviewer 2: Clarification: this new feature's purpose is to "allow the initialization of collection-type properties (even if they are read-only) using element attribute syntax". The CreateFromString attribute is one potential way in which a collection type might take advantage of this feature.

Author: Fixed.

<!-- Use this section to provide background context for the new API(s) 
in this spec. -->

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

# Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->
## Overview
A new XAML language feature that allows the initialization of collection-type properties (even if they are read-only) using element attribute syntax.

## Implementation Details

> Read-only collection-type properties will be assigned a new CreateFromString attribute. The CreateFromString attribute will allow them to be defined/created from a comma-delimited string. In terms of Grid, this means that the ColumnDefinitions and RowDefinitions properties are able to be defined by providing a comma-delimited string of ColumnDefinitions (Width values) and a comma-delimited string of RowDefinitions (Height values), respectively. However, this feature is also applicable to other scenarios within the XAML language where collection-typed read-only properties are created (see Remarks for examples of such scenarios). 

Reviewer 2: At a high level, what we're implementing is the following:
1. When the parser encounters an element attribute that is assigning to a collection-type property, it recognizes that initialization of a collection from a string is being requested.
2. The parser tokenizes the attribute value (a comma-delimited string) into a series of substrings, each of which represents the initialization value of a single element of the collection
3. The parser will perform the usual "initialize an instance of a type from a string" operation. 
4. If the type has a ContentProperty, then a new instance of the type will be created (zero-argument constructor) and the string value will be assigned to the content property (this might in turn invoke a type converter, but the details are at a lower abstraction level and so are not germane to this discussion).
5. If the type has either a built-in type converter or a CreateFromString attribute specifying a factory method, then that will be invoked with the string value as an argument, returning a new instance of the type.
6. If neither step 4 nor 5 is performed, then this is an error.

For Grid-specific portion of this work, we'll be following step 4 from above, adding the ContentProperty attribute to RowDefinition and ColumnDefinition.

Author: Hope its okay that I used your words in the spec - added this in.

# Examples
### New succinct syntax
The code below has the same functionality as the code shown above with the original syntax. It creates a grid and defines five different rows and columns, each with their own specific height/width, and adds them to the Grid. 
```xml
<Grid ColumnDefinitions="1*, 2*, Auto, *, 300"
      RowDefinitions="1*, Auto, 25, 14, 20">
</Grid>
```

### New succinct syntax for more complex types
The code below shows an example of how the syntax would be used for collection-typed properties that have more complex value types.

This shows how Point objects will be delimited by a comma, but indvidual values will be encased in single quotes.
```xml
<Foo PointList=" '1,2', '9,4', '2,6'"/>
```

> This shows how String values will be supported, even if they include commas or single quotes inside of them. Each string will be encased in single quotes and delimited by commas, and single quotes inside of a string will be escaped by using an HTML entity. The collection shown below has 4 elements, and 'hello, world!' is one element.

Reviewer 2: Following the XML grammar, if the attribute value is surrounded by `'`, then `"` is the grouping character that is used within the string. Basically, attribute values can be enclosed with either `'` or `"`, which then allows the other quotation mark to be used within the value itself without needing to be escaped with an XML character entity.

# Remarks
<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->
### Syntax Details and Corner Cases
The new succinct syntax is formed by assigning a read-only collection-typed property to a collection of values. This collection of values is encased between a set of double quotes. String values, as well as any other more complex type that may require commas within it (i.e. Point objects, sets), are encased between a set of single quotes. This ensures that string values may contain commas, and Points, sets, or any other kind of nested collection object may be supported. To use a single quote within a complex value that is encased in single quotes (i.e., to put a single quote or apostrophe inside of a string value), the HTML entity (&apos;) should be used to escape that character.

### Other Use Cases for Succinct Syntax

> The succinct syntax was created with the goal of making Grid more intuitive. However, this new syntax will be implemented as a language feature and work properly for any read-only collection-typed property. The current syntaxes will still be functional for all other scenarios, as it will be with Grid. Examples of some of these scenarios include, but are not limited to:

Reviewer 2: The language feature will work for any collection-typed property, not only those that are read-only.

          <ColumnDefinition Width="*" />
          <ColumnDefinition Width="300" />
    </Grid.ColumnDefinitions>
    <Grid.RowDefinitions>
          <RowDefinition Height="1*" />
          <RowDefinition Height="Auto" />
          <RowDefinition Height="25" />
          <RowDefinition Height= "14" />
          <RowDefinition Height="20" />
    </Grid.RowDefinitions>
</Grid>
```
a duplicate property error will be thrown. This is already implemented for read/write properties, but will be implemented for read-only properties such as these as well. 

<!-- Option 1: Give a one or two line description of each API (type
and member), or at least the ones that aren't obvious
from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

# API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

**ColumnDefinition API**
```csharp

> [contract(Windows.Foundation.UniversalApiContract, 1)]

Reviewer 2: I don't think this is correct. We should follow up with @MikeHillberg off-line about our contract story.

Reviewer: This is on the Width property, which is indeed UAC1.

Author: In WinUI3, we won't have the UAC contract

[webhosthidden]

/// Sets the ColumnDefinition's content property to the Width property.

> [contentproperty("Width")]

Reviewer 2: It'd be cool if we had some way to highlight this to indicate that it's what's being added.

Author: I tried adding in html/even making it a block quote, but nothing worked because its inside of a code snippet

Reviewer 2: That's unfortunate. :(

Reviewer: This is a frustrating issue. Maybe we should adopt a pattern of highlighting lines with comments. E.g.

```c#
    Double SomeProperty;

    // vvvvv
    int TheNewProperty;
    // ^^^^^
    string SomeOtherProperty
```

Author: I like that idea, I'll incorporate into this spec for now!

[static_name("Windows.UI.Xaml.Controls.IColumnDefinitionStatics", 06b0d728-d044-40c6-942e-ae60eac74851)]
[interface_name("Windows.UI.Xaml.Controls.IColumnDefinition", f7f1b229-f024-467f-970a-7e705615db7b)]
runtimeclass ColumnDefinition : Windows.UI.Xaml.DependencyObject
{
    ColumnDefinition();
    Windows.UI.Xaml.GridLength Width;
    Double MaxWidth;
    Double MinWidth;
    Double ActualWidth{ get; };
    static Windows.UI.Xaml.DependencyProperty WidthProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaxWidthProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MinWidthProperty{ get; };
};
```

**RowDefinition API**
```csharp
[contract(Windows.Foundation.UniversalApiContract, 1)]
[webhosthidden]

/// Sets the RowDefinition's content property to the Height property.

> [contentproperty("Height")]

Reviewer 2: It'd be cool if we had some way to highlight this to indicate that it's what's being added.

[static_name("Windows.UI.Xaml.Controls.IRowDefinitionStatics", 5adf3fe5-2056-4724-94d6-e4812b022ec8)]
[interface_name("Windows.UI.Xaml.Controls.IRowDefinition", 4abae829-d80c-4a5e-a48c-f8b3d3b6533d)]
runtimeclass RowDefinition : Windows.UI.Xaml.DependencyObject
{
    RowDefinition();
    Windows.UI.Xaml.GridLength Height;
    Double MaxHeight;
    Double MinHeight;
    Double ActualHeight{ get; };
    static Windows.UI.Xaml.DependencyProperty HeightProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaxHeightProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MinHeightProperty{ get; };
};
```

**ColumnDefinitions API**
```csharp
[contract(Windows.Foundation.UniversalApiContract, 1)]
[webhosthidden]

/// Allows ColumnDefinitions to be createable from string using the method Create.

> [CreateFromString(MethodName = nameof(Create))]

Reviewer 2: We're not employing the CreateFromString attribute mechanism.

runtimeclass ColumnDefinitionCollection :[default] Windows.Foundation.Collections.IVector<Windows.UI.Xaml.Controls.ColumnDefinition>
{
    [method_name("Create")] ColumnDefinitionCollection Create(string Widths);
};
```

**RowDefinitions API**
```csharp
[contract(Windows.Foundation.UniversalApiContract, 1)]
[webhosthidden]

/// Allows RowDefinitions to be createable from string using the method Create.

> [CreateFromString(MethodName = nameof(Create))]

Reviewer 2: We're not employing the CreateFromString attribute mechanism.
